File: scripts/importar-excel.js
const XLSX = require('xlsx');
const db = require('../database');
const path = require('path');
const fs = require('fs');

function importarDesdeExcel(excelPath) {
    console.log(`\nLeyendo archivo Excel: ${excelPath}`);
    
    if (!fs.existsSync(excelPath)) {
        throw new Error(`El archivo ${excelPath} no existe`);
    }

    const workbook = XLSX.readFile(excelPath);
    const sheetName = workbook.SheetNames[0];
    const worksheet = workbook.Sheets[sheetName];
    const data = XLSX.utils.sheet_to_json(worksheet);

    console.log(`Filas encontradas: ${data.length}`);

    // Identificar responsables únicos
    console.log('\nIdentificando responsables...');
    const responsablesSet = new Set();
    
    data.forEach(row => {
        const responsable = row['Responsable'];
        if (responsable && responsable !== 'Responsable' && responsable.trim()) {
            responsablesSet.add(String(responsable).trim());
        }
    });

    console.log(`Responsables encontrados: ${Array.from(responsablesSet).join(', ')}`);

    // Crear responsables en la base de datos
    const responsablesCreados = {};
    responsablesSet.forEach(nombre => {
        const id = db.agregarResponsable(nombre);
        responsablesCreados[nombre] = id;
    });

    console.log(`✓ ${Object.keys(responsablesCreados).length} responsables creados/actualizados`);

    // Importar clientes
    console.log('\nImportando clientes...');
    let clientesImportados = 0;
    let errores = 0;

    data.forEach((row, index) => {
        try {
            const nombre = row['Nombre cliente'];
            if (!nombre || nombre === 'Nombre cliente') {
                return; // Saltar filas vacías o encabezados
            }

            const direccion = row['Dirección'] ? String(row['Dirección']).trim() : null;
            const comuna = row['Comuna'] ? String(row['Comuna']).trim() : null;
            const celular = row['Celular'] ? String(row['Celular']).trim() : null;
            const responsableVal = row['Responsable'];
            const diaAtencion = row['día de atención'] ? String(row['día de atención']).trim() : null;
            const precioVal = row['precio'];

            // Obtener responsable_id
            let responsableId = null;
            if (responsableVal && responsableVal !== 'Responsable') {
                const responsableName = String(responsableVal).trim();
                if (responsablesCreados[responsableName]) {
                    responsableId = responsablesCreados[responsableName];
                }
            }

            // Procesar precio
            let precioPorVisita = 0;
            if (precioVal !== undefined && precioVal !== null) {
                try {
                    if (typeof precioVal === 'number') {
                        precioPorVisita = precioVal;
                    } else {
                        const precioStr = String(precioVal).replace(/[$,\s]/g, '').trim();
                        precioPorVisita = parseFloat(precioStr) || 0;
                    }
                } catch (e) {
                    precioPorVisita = 0;
                }
            }

            // Insertar cliente
            db.agregarCliente(
                String(nombre).trim(),
                direccion,
                comuna,
                celular,
                responsableId,
                diaAtencion,
                precioPorVisita
            );

            clientesImportados++;
            if (clientesImportados % 50 === 0) {
                console.log(`  Importados ${clientesImportados} clientes...`);
            }
        } catch (error) {
            errores++;
            console.error(`  Error en fila ${index + 2}: ${error.message}`);
        }
    });

    console.log(`\n✓ Importación completada!`);
    console.log(`  - Responsables: ${Object.keys(responsablesCreados).length}`);
    console.log(`  - Clientes importados: ${clientesImportados}`);
    if (errores > 0) {
        console.log(`  - Errores: ${errores}`);
    }

    return { clientesImportados, responsables: Object.keys(responsablesCreados).length };
}

// Ejecutar si se llama directamente
if (require.main === module) {
    const excelPath = process.argv[2] || 'Base de Datos United al 28 oct 2025.xlsx';
    
    try {
        importarDesdeExcel(excelPath);
        process.exit(0);
    } catch (error) {
        console.error(`\n❌ Error durante la importación: ${error.message}`);
        process.exit(1);
    }
}

module.exports = importarDesdeExcel;

File: importar_excel.py
"""
Script para importar datos del Excel a la base de datos SQLite.
"""
import openpyxl
from database import Database
import sys


def importar_desde_excel(excel_path: str, db_path: str = "piscinas.db"):
    """Importa los datos del Excel a la base de datos."""
    db = Database(db_path)
    
    print(f"Leyendo archivo Excel: {excel_path}")
    wb = openpyxl.load_workbook(excel_path)
    ws = wb.active
    
    # Leer encabezados
    headers = {}
    for col in range(1, ws.max_column + 1):
        val = ws.cell(1, col).value
        if val:
            headers[val] = col
    
    print(f"Columnas encontradas: {list(headers.keys())}")
    
    # Procesar filas
    clientes_importados = 0
    responsables_creados = {}
    
    # Primero, crear responsables únicos
    print("\nIdentificando responsables...")
    responsables_set = set()
    for row in range(2, ws.max_row + 1):
        responsable_val = ws.cell(row, headers.get('Responsable', 0)).value
        if responsable_val and responsable_val not in ['Responsable', None]:
            responsables_set.add(str(responsable_val).strip())
    
    print(f"Responsables encontrados: {list(responsables_set)}")
    
    # Crear responsables en la base de datos
    for resp_name in responsables_set:
        resp_id = db.agregar_responsable(resp_name)
        responsables_creados[resp_name] = resp_id
    
    # Importar clientes
    print("\nImportando clientes...")
    for row in range(2, ws.max_row + 1):
        # Obtener datos del cliente
        nombre = ws.cell(row, headers.get('Nombre cliente', 0)).value
        if not nombre or nombre == 'Nombre cliente':
            continue
        
        direccion = ws.cell(row, headers.get('Dirección', 0)).value
        comuna = ws.cell(row, headers.get('Comuna', 0)).value
        celular = ws.cell(row, headers.get('Celular', 0)).value
        responsable_val = ws.cell(row, headers.get('Responsable', 0)).value
        dia_atencion = ws.cell(row, headers.get('día de atención', 0)).value
        precio_val = ws.cell(row, headers.get('precio', 0)).value
        
        # Limpiar y procesar datos
        nombre = str(nombre).strip()
        direccion = str(direccion).strip() if direccion else None
        comuna = str(comuna).strip() if comuna else None
        celular = str(celular).strip() if celular else None
        dia_atencion = str(dia_atencion).strip() if dia_atencion else None
        
        # Obtener responsable_id
        responsable_id = None
        if responsable_val and responsable_val not in ['Responsable', None]:
            responsable_name = str(responsable_val).strip()
            if responsable_name in responsables_creados:
                responsable_id = responsables_creados[responsable_name]
        
        # Procesar precio
        precio_por_visita = 0
        if precio_val:
            try:
                # Intentar convertir a número
                if isinstance(precio_val, (int, float)):
                    precio_por_visita = float(precio_val)
                else:
                    precio_str = str(precio_val).replace('$', '').replace(',', '').strip()
                    precio_por_visita = float(precio_str)
            except (ValueError, TypeError):
                precio_por_visita = 0
        
        # Insertar cliente
        try:
            cliente_id = db.agregar_cliente(
                nombre=nombre,
                direccion=direccion,
                comuna=comuna,
                celular=celular,
                responsable_id=responsable_id,
                dia_atencion=dia_atencion,
                precio_por_visita=precio_por_visita
            )
            clientes_importados += 1
            if clientes_importados % 50 == 0:
                print(f"  Importados {clientes_importados} clientes...")
        except Exception as e:
            print(f"  Error al importar cliente {nombre}: {e}")
            continue
    
    print(f"\n✓ Importación completada!")
    print(f"  - Responsables creados: {len(responsables_creados)}")
    print(f"  - Clientees importados: {clientes_importados}")
    
    return clientes_importados


if __name__ == "__main__":
    excel_path = "Base de Datos United al 28 oct 2025.xlsx"
    
    if len(sys.argv) > 1:
        excel_path = sys.argv[1]
    
    try:
        importar_desde_excel(excel_path)
    except FileNotFoundError:
        print(f"Error: No se encontró el archivo {excel_path}")
        sys.exit(1)
    except Exception as e:
        print(f"Error durante la importación: {e}")
        import traceback
        traceback.print_exc()
        sys.exit(1)

